twoliter/src/lock.rs

```diff
@@ -139,6 +249,124 @@ impl Lock {
         Ok(lock)
     }
 
+    fn external_kit_metadata(&self) -> ExternalKitMetadata {
+        ExternalKitMetadata {
+            sdk: self.sdk.clone(),
+            kits: self.kit.clone(),
+        }
+    }
+
+    /// Fetches all external kits defined in a Twoliter.lock to the build directory
+    pub(crate) async fn fetch(&self, project: &Project, arch: &str) -> Result<()> {
+        let target_dir = project.external_kits_dir();
+        create_dir_all(&target_dir)
+            .await
+            .context("failed to create external-kits directory")?;
+        for image in self.kit.iter() {
+            self.extract_kit(&target_dir, image, arch).await?;
+        }
+        let mut kit_list = Vec::new();
+        let mut ser =
+            serde_json::Serializer::with_formatter(&mut kit_list, CanonicalJsonFormatter::new());
+        self.external_kit_metadata()
+            .serialize(&mut ser)
+            .context("failed to serialize external kit metadata")?;
+        write(project.external_kits_metadata(), kit_list.as_slice())
+            .await
+            .context("failed to write external kit metadata")?;
+
+        Ok(())
+    }
+
+    async fn get_manifest(&self, image: &LockedImage, arch: &str) -> Result<ManifestView> {
+        let manifest_bytes = docker!(
+            ["manifest", "inspect", image.source.as_str()],
+            format!("failed to find a kit {}", image.to_string())
+        );
+        let manifest_list: ManifestListView = serde_json::from_slice(manifest_bytes.as_slice())
+            .context("failed to deserialize manifest list")?;
+        let docker_arch = Architecture::try_from(arch)?;
+        manifest_list
+            .manifests
+            .iter()
+            .find(|x| x.platform.as_ref().unwrap().architecture == docker_arch)
+            .cloned()
+            .context(format!(
+                "could not find kit image for architecture '{}' at {}",
+                docker_arch, image.source
+            ))
+    }
+
+    async fn pull_image<P>(&self, path: P, digest_uri: &str) -> Result<IndexView>
+    where
+        P: AsRef<Path>,
+    {
+        let oci_archive = docker!(
+            ["save", digest_uri],
+            format!("failed to fetch kit and save to disk from {}", digest_uri)
+        );
```

Comment:
Is this saving the entire image into memory via console output?

If it's implicitly saving a file, let's make that explicit. If it's not, please make it save a file. We don't want to read a remote file into a stack variable ...